cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/app/etl/batch/ETLMapReduce.java

```diff
@@ -355,8 +459,62 @@ public void write(IN input) throws Exception {
           context.write(outputName, outputRecord.getKey(), outputRecord.getValue());
         }
       }
-      // TODO: write errors to error dataset
+
+      if (errorDatasetName != null && !emitter.getErrors().isEmpty()) {
+        for (InvalidEntry entry : emitter.getErrors()) {
+          context.write(errorDatasetName, new AvroKey<>(getGenericRecordForInvalidEntry(entry)), NullWritable.get());
+        }
+      }
+
       emitter.reset();
     }
   }
+
+  private static GenericRecord getGenericRecordForInvalidEntry(InvalidEntry invalidEntry) {
+    org.apache.avro.Schema avroSchema = new org.apache.avro.Schema.Parser().parse(ERROR_SCHEMA.toString());
+    GenericRecordBuilder recordBuilder = new GenericRecordBuilder(avroSchema);
+    recordBuilder.set(Constants.ErrorDataset.ERRCODE, invalidEntry.getErrorCode());
+    recordBuilder.set(Constants.ErrorDataset.ERRMSG, invalidEntry.getErrorMsg());
+
+    String errorMsg;
+    if (invalidEntry.getInvalidRecord() instanceof StructuredRecord) {
+      StructuredRecord record = (StructuredRecord) invalidEntry.getInvalidRecord();
+      try {
+        errorMsg = StructuredRecordStringConverter.toJsonString(record);
+      } catch (IOException e) {
+        errorMsg =   "Exception while converting StructuredRecord to String, " + e.getCause();
+      }
+    } else {
+      errorMsg =  String.format("Error Entry is of type %s, only records of type " +
+                                  StructuredRecord.class.getName() +
+                                  "is supported currently", invalidEntry.getInvalidRecord().getClass().getName());
```

Comment:
nit: no need for concatenation if you're already using ``String.format``, right?